File: benchmarks/ts-bench/src/tests/size-matrix.ts
import type { Config } from '../config.js';
import type { TierResult } from '../report.js';
import * as protocol from '../protocol.js';
import { printHeader, fmtCount, fmtBytesPerSec } from '../stats.js';
import { generateBenchToken } from '../jwt.js';
import { tiersFor } from '../config.js';

const PAYLOAD_SIZES = [64, 256, 1024, 4096, 16384, 65536];
const CELL_DURATION_SECS = 5;

export async function run(config: Config): Promise<TierResult[]> {
  printHeader(
    'TEST 4: Payload Size x Connections Matrix',
    `${PAYLOAD_SIZES.length} sizes x N tiers, ${CELL_DURATION_SECS}s per cell.`
  );

  const tiers = tiersFor(config, 'size-matrix');
  const token = generateBenchToken(Buffer.from(config.secret), 'bench-user');
  const results: TierResult[] = [];

  // Build payloads
  const payloads = PAYLOAD_SIZES.map((size) => {
    const payload = protocol.buildPayload(size);
    return { size, payload, buf: Buffer.from(payload) };
  });

  // Print header
  const tierLabels = tiers.map((t) => (t >= 1000 ? `${(t / 1000).toFixed(0)}K` : `${t}`));
  console.log(`\n  ### Messages per second\n`);
  console.log(`  | Size \\ Conns | ${tierLabels.join(' | ')} |`);
  console.log(`  |${'-'.repeat(13)}|${tierLabels.map(() => '------').join('|')}|`);

  for (const { size, buf } of payloads) {
    const sizeLabel =
      size >= 1024 ? `${(size / 1024).toFixed(0)} KB` : `${size} B`;
    const cells: string[] = [];

    for (const n of tiers) {
      const connections = await protocol.connectBatch(
        config.host, config.port, token, n, config.batchSize,
        'compression=false&protocol_version=1'
      );

      if (connections.length === 0) {
        cells.push('-');
        continue;
      }

      const actual = connections.length;
      const deadline = Date.now() + CELL_DURATION_SECS * 1000;
      const totalStart = process.hrtime.bigint();

      const sendPromises = connections.map((ws) => matrixSendLoop(ws, buf, deadline));
      const sendResults = await Promise.allSettled(sendPromises);

      const totalSecs = Number(process.hrtime.bigint() - totalStart) / 1_000_000_000;

      let totalMessages = 0;
      let totalBytes = 0;
      const wsList: protocol.WsConnection[] = [];

      for (const r of sendResults) {
        if (r.status === 'fulfilled') {
          totalMessages += r.value.count;
          totalBytes += r.value.bytes;
          wsList.push(r.value.ws);
        }
      }

      const msgsPerSec = totalMessages / totalSecs;
      cells.push(fmtCount(Math.round(msgsPerSec)));

      results.push({
        tier: n,
        connected: actual,
        errors: n - actual,
        durationSecs: totalSecs,
        messagesSent: totalMessages,
        messagesReceived: 0,
        bytesSent: totalBytes,
        latency: null,
        extra: { payload_size: size, msgs_per_sec: msgsPerSec },
      });

      await protocol.closeAll(wsList);
    }

    console.log(`  | **${sizeLabel}** | ${cells.join(' | ')} |`);
  }

  return results;
}

async function matrixSendLoop(
  ws: protocol.WsConnection,
  msgBuf: Buffer,
  deadline: number,
): Promise<{ count: number; bytes: number; ws: protocol.WsConnection }> {
  let count = 0;
  let bytes = 0;
  const msgSize = msgBuf.length;

  return new Promise((resolve) => {
    function send() {
      while (Date.now() < deadline) {
        if ((ws as any).bufferedAmount > 16 * 1024 * 1024) {
          setTimeout(send, 1);
          return;
        }
        ws.send(msgBuf, { binary: false }, (err) => {
          if (err) resolve({ count, bytes, ws });
        });
        count++;
        bytes += msgSize;
        if (count % 1000 === 0) {
          setImmediate(send);
          return;
        }
      }
      resolve({ count, bytes, ws });
    }
    send();
  });
}

File: benchmarks/rust-bench/src/tests/size_matrix.rs
use crate::config::{Cli, MATRIX_SIZES};
use crate::protocol::{self, WsStream};
use crate::report::{self, TierResult};
use crate::stats;
use futures_util::SinkExt;
use std::collections::HashMap;
use std::sync::atomic::{AtomicU64, Ordering};
use std::sync::Arc;
use tokio_tungstenite::tungstenite::protocol::Message;

/// Test 4: Payload Size x Connections Matrix
/// Full throughput matrix across payload sizes and connection counts.
pub async fn run(cli: &Cli) -> Vec<TierResult> {
    stats::print_header(
        "TEST 4: Payload Size x Connections Matrix",
        "Throughput (msg/s) for each (size, connections) pair. 5s burst each.",
    );

    let tiers = cli.tiers_for(crate::config::TestName::SizeMatrix);
    let token = crate::jwt::generate_bench_token(cli.secret.as_bytes(), "bench-user");
    let sizes = MATRIX_SIZES;
    let burst_secs = 5u64; // Shorter than full test for matrix cells

    let mut results = Vec::new();
    let mut matrix: HashMap<(usize, usize), TierResult> = HashMap::new();

    for &size in sizes {
        let payload = protocol::build_payload(size);
        let actual_size = payload.len();
        println!(
            "\n  === Payload: {} ({} bytes actual) ===",
            format_size(size),
            actual_size
        );

        for &n in &tiers {
            eprint!("    {} conns: ", n);

            let connections = protocol::connect_batch(
                &cli.host,
                cli.port,
                &token,
                n,
                cli.batch_size,
                "compression=false&protocol_version=1",
            )
            .await;

            if connections.is_empty() {
                eprintln!("FAILED");
                continue;
            }

            let actual = connections.len();
            let total_sent = Arc::new(AtomicU64::new(0));
            let total_bytes = Arc::new(AtomicU64::new(0));
            let duration = std::time::Duration::from_secs(burst_secs);

            let start = tokio::time::Instant::now();
            let mut handles = Vec::with_capacity(actual);

            for ws in connections {
                let sent = total_sent.clone();
                let bytes = total_bytes.clone();
                let msg = payload.clone();
                let plen = actual_size as u64;

                handles.push(tokio::spawn(async move {
                    matrix_send_loop(ws, &msg, plen, duration, sent, bytes).await
                }));
            }

            let ws_list: Vec<WsStream> = futures_util::future::join_all(handles)
                .await
                .into_iter()
                .filter_map(|r| r.ok())
                .collect();

            let elapsed = start.elapsed().as_secs_f64();
            let sent = total_sent.load(Ordering::Relaxed);
            let bytes = total_bytes.load(Ordering::Relaxed);
            let msg_s = sent as f64 / elapsed;
            let mb_s = bytes as f64 / elapsed / 1_000_000.0;

            eprintln!("{} msg/s, {:.1} MB/s", stats::fmt_rate(msg_s), mb_s);

            let tier_result = TierResult {
                tier: n,
                connected: actual,
                errors: n - actual,
                duration_secs: elapsed,
                messages_sent: sent,
                messages_received: 0,
                bytes_sent: bytes,
                latency: None,
                extra: serde_json::json!({
                    "payload_size": size,
                    "actual_payload_size": actual_size,
                    "msg_per_sec": msg_s,
                    "mb_per_sec": mb_s,
                }),
            };

            matrix.insert((size, n), tier_result.clone());
            results.push(tier_result);

            protocol::close_all(ws_list).await;

            // Brief pause between cells
            tokio::time::sleep(std::time::Duration::from_millis(500)).await;
        }
    }

    // Print matrix table
    report::print_matrix_table(sizes, &tiers, &matrix);

    results
}

async fn matrix_send_loop(
    mut ws: WsStream,
    payload: &str,
    payload_len: u64,
    duration: std::time::Duration,
    total_sent: Arc<AtomicU64>,
    total_bytes: Arc<AtomicU64>,
) -> WsStream {
    let deadline = tokio::time::Instant::now() + duration;
    let mut local_sent = 0u64;
    let mut local_bytes = 0u64;
    let msg = Message::Text(payload.to_string().into());

    while tokio::time::Instant::now() < deadline {
        if ws.send(msg.clone()).await.is_err() {
            break;
        }
        local_sent += 1;
        local_bytes += payload_len;

        if local_sent.is_multiple_of(64) {
            let _ = ws.flush().await;
        }
    }

    total_sent.fetch_add(local_sent, Ordering::Relaxed);
    total_bytes.fetch_add(local_bytes, Ordering::Relaxed);
    ws
}

fn format_size(n: usize) -> String {
    if n >= 1_048_576 {
        format!("{} MB", n / 1_048_576)
    } else if n >= 1_024 {
        format!("{} KB", n / 1_024)
    } else {
        format!("{} B", n)
    }
}
